pkg/storegateway/bucket_index_reader.go

```diff
@@ -268,6 +224,98 @@ func (r *bucketIndexReader) expandedPostings(ctx context.Context, ms []*labels.M
 	return ps, nil
 }
 
+// toPostingGroups returns a set of labels for which to look up postings lists. It guarantees that
+// each postingGroup's keys exist in the index. The order of the returned labels
+// is the same as iterating through the posting groups and for each group adding all keys.
+func toPostingGroups(ms []*labels.Matcher, indexhdr indexheader.Reader) ([]postingGroup, []labels.Label, error) {
+	var (
+		rawPostingGroups = make([]rawPostingGroup, 0, len(ms))
+		allRequested     = false
+		hasAdds          = false
+	)
+
+	// NOTE: Derived from tsdb.PostingsForMatchers.
+	for _, m := range ms {
+		// Each group is separate to tell later what postings are intersecting with what.
+		rawPostingGroups = append(rawPostingGroups, toRawPostingGroup(m))
+	}
+
+	// We can check whether their requested values exist
+	// in the index. We do these checks in a specific order, so we minimize the reads from disk
+	// and/or to minimize the number of label values we help in memory (assumption being that indexhdr.LabelValues
+	// is likely to return a lot of values).
+	sort.Slice(rawPostingGroups, func(i, j int) bool {
+		// First we check the non-lazy groups, since for those we only need to call PostingsOffset, which
+		// is less expensive than LabelValues for the lazy groups.
+		ri, rj := rawPostingGroups[i], rawPostingGroups[j]
+		if ri.isLazy != rj.isLazy {
+			return !ri.isLazy
+		}
+
+		// Within the lazy/non-lazy groups we sort by the number of keys they have.
+		// The idea is that groups with fewer keys are more likely to match no actual series.
+		if len(ri.keys) != len(rj.keys) {
+			return len(ri.keys) < len(rj.keys)
+		}
+
+		// Sort by label name to make this a deterministic sort.
+		return ri.labelName < rj.labelName
+	})
+
+	filteredPostingGroups := make([]postingGroup, 0, len(rawPostingGroups))
+	// Next we check whether the posting groups won't select an empty set of postings.
+	// We start with the ones that have a known set of values because it's less expensive to check them in
+	// the index header.
```

Comment:
[nit] 
```suggestion
	// Based on the previous sorting, we start with the ones that have a known set of values because it's less expensive to check them in
	// the index header.
```